phasta/phCGNSgbc.cc

```diff
@@ -0,0 +1,460 @@
+{
+  int nelem = o.blocks.boundary.nElements[block];
+// CGNS wants surface elements  int nvert = o.blocks.boundary.keys[block].nElementVertices;
+  int nvert = o.blocks.boundary.keys[block].nBoundaryFaceEdges;
+  //c.setSize(nelem * nvert);
+  size_t i = 0;
+  for (int elem = 0; elem < nelem; ++elem)
+    for (int vert = 0; vert < nvert; ++vert)
+      c[i++] = o.arrays.ncorp[o.arrays.ienb[block][elem][vert]];
+  PCU_ALWAYS_ASSERT(i == nelem*nvert);
+}
+
+void getInterfaceConnectivityCGNS // not extended yet other than transpose
+(
+  Output& o,
+  int block,
+  apf::DynamicArray<int>& c
+)
+{
+  int nelem = o.blocks.interface.nElements[block];
+  int nvert0 = o.blocks.interface.keys[block].nElementVertices;
+  int nvert1 = o.blocks.interface.keys[block].nElementVertices1;
+  c.setSize(nelem * (nvert0 + nvert1));
+  size_t i = 0;
+  for (int elem = 0; elem < nelem; ++elem)
+    for (int vert = 0; vert < nvert0; ++vert)
+      c[i++] = o.arrays.ncorp[o.arrays.ienif0[block][elem][vert]-1];
+  for (int elem = 0; elem < nelem; ++elem)
+    for (int vert = 0; vert < nvert1; ++vert)
+      c[i++] = o.arrays.ncorp[o.arrays.ienif1[block][elem][vert]-1];
+  PCU_ALWAYS_ASSERT(i == c.getSize());
+}
+
+// renamed but not updated yet
+void getNaturalBCCodesCGNS(Output& o, int block, int* codes)
+{
+  int nelem = o.blocks.boundary.nElements[block];
+  size_t i = 0;
+  for (int elem = 0; elem < nelem; ++elem)
+      codes[i++] = o.arrays.ibcb[block][elem][1]; //srfID is the second number so 1
+// if we wanted we could use PHASTA's bit in coding in the first number to us attributes to set
+// arbitrary combinations of BCs but leaving that out for now
+}
+
+// renamed and calling the renamed functions above with output writes commented as they are PHASTA file style
+void writeBlocksCGNS(int F,int B,int Z, Output& o)
+{
+  int params[MAX_PARAMS];
+  int E;
+  cgsize_t e_owned, e_start,e_end;
+  cgsize_t e_startg,e_endg;
+  cgsize_t e_written=0;
+  for (int i = 0; i < o.blocks.interior.getSize(); ++i) {
+    BlockKey& k = o.blocks.interior.keys[i];
+    std::string phrase = getBlockKeyPhrase(k, "connectivity interior ");
+    params[0] = o.blocks.interior.nElements[i];
+    e_owned = o.blocks.interior.nElements[i];
+    int nvert = o.blocks.interior.keys[i].nElementVertices;
+    cgsize_t* e = (cgsize_t *)malloc(nvert * e_owned * sizeof(cgsize_t));
+    getInteriorConnectivityCGNS(o, i, e);
+    /* create data node for elements */
+    e_startg=1+e_written; // start for the elements of this topology
+    e_endg=e_written + PCU_Add_Long(e_owned); // end for the elements of this topology
+    switch(nvert){
+      case 4:
+        if (cgp_section_write(F, B, Z, "Tet", CG_TETRA_4, e_startg, e_endg, 0, &E))
+           cgp_error_exit();
+        break;
+      case 5:
+    free(e);
+        if (cgp_section_write(F, B, Z, "Pyr", CG_PYRA_5, e_startg, e_endg, 0, &E))
+           cgp_error_exit();
+        break;
+      case 6:
```

Comment:
```suggestion
        break;
      case 5:
        if (cgp_section_write(F, B, Z, "Pyr", CG_PYRA_5, e_startg, e_endg, 0, &E))
           cgp_error_exit();
        break;
      case 6:
```

Pretty sure this `free(e)` is wrong. Or at least I'm quite curious as to why it's needed for pyramids, but no one else?

Reply:
Thanks for catching this spurious paste.